### ISSUE TITLE
ReadOnlyObservableCollection.CollectionChanged should be made public

### ISSUE THREAD

**Issue by davkean** [dotnet/corefx#970]
The `ReadOnlyObservableCollection.CollectionChanged` event is protected, which is flawed, it prevents you from being able to observe changes to it without casting to `INotifyCollectionChanged`. 

I suspect the original thinking was because the collection is "read-only" it never changes - this is not correct. _Read-only_ in .NET means that the consumer has a read-only view over a collection (which is why we project `IReadOnlyList<T>` to `IVectorView<T>`), not that the contents are immutable and never change.

While we've made `protected` members `public` in the past (such as `Exception.HResult`), this event was marked `virtual`, which while binary compatible, would be a source-breaking change for those that overrode it. I suspect very little have overridden it, but we may want to consider adding a new event with a different name such as `Changed`.

**Comment by sharwell** [comment]
I believe the original thinking was the `NotifyCollectionChangedEventArgs` structure could expose the underlying (non-readonly) list in certain scenarios. You can see evidence of this [in the reference source code](http://referencesource.microsoft.com/#System/compmod/system/collections/specialized/notifycollectionchangedeventargs.cs,322). Exposing the underlying collection would require the user to pass the underlying collection (or perhaps an `ArraySegment<T>` wrapping the underlying collection) to one of a `NotifyCollectionChangedEventArgs` constructors [like this one](http://referencesource.microsoft.com/#System/compmod/system/collections/specialized/notifycollectionchangedeventargs.cs,221). The reference source code does not contain any example of this, but it's conceivable that a user may have implemented the "Reset" event as a "Replace All Items" event.

The code which would be responsible for the above scenario appears to be removed from this project. In addition, I agree that it's unlikely users would override this event. The primary reason to override it would be to address the specific accessibility problem you describe here. A more likely scenario would be implementing a new publicly-exposed event which simply attaches listeners to the event in the base class. The latter use case would not break if the accessibility of the event was changed.

**Comment by davkean** [comment]
> I believe the original thinking was the NotifyCollectionChangedEventArgs structure could expose the underlying (non-readonly) list in certain scenarios

That was the behavior on Compact Framework, but not the behavior on full CLR. Given that CF implementation of this particular class came _after_ the desktop implementation, I don't believe this is the history.

**Comment by sharwell** [comment]
I agree that it's unlikely users would override this event. The primary reason to override it would be to address the specific accessibility problem you describe here, but that isn't allowed in C#. A more likely scenario would be implementing a new publicly-exposed event which simply attaches listeners to the event in the base class. The latter use case would not break if the accessibility of the event was changed.

**Comment by davkean** [comment]
@nguerrera @terrajobst, thoughts?

**Comment by nguerrera** [comment]
I agree with this proposal. We'd have to change desktop and source-breaking can mean app-breaking in ASP.NET, correct? I think the risk is very low, but it would depend on what we're willing to accept in the desktop framework I suppose. It would be unfortunate to clutter things with a property name that doesn't match the interface. cc @ericstj for compat concern.

**Comment by jnm2** [comment]
+1 Please do this! As it stands there is no difference between `ReadOnlyObservableCollection` and `ReadOnlyCollection` unless you cast which is not API-friendly. The sole reason for the existence of `ReadOnlyObservableCollection` is to add events to `ReadOnlyCollection`.

**Comment by nguerrera** [comment]
See my comnent on #1029. This is actually binary breaking.

**Comment by terrajobst** [comment]
Can we add an event under a new name, such as just `Changed`? We could even go the extra mile and make the event args strongly typed i.e. provide the values as `IList<T>` rather than the non-generic `IList`.

**Comment by danmosemsft** [comment]
Anyone willing to write up a formal [api proposal](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md) ?

**Comment by AtosNicoS** [comment]
Any updates on this issue?
